tests/beta/tapes/test_rev.py

```diff
@@ -0,0 +1,601 @@
+        """Tests that the correct gradient is computed for qnodes which
+        use the same parameter in multiple gates."""
+
+        from gate_data import Rotx as Rx, Roty as Ry, Rotz as Rz
+
+        def expZ(state):
+            return np.abs(state[0]) ** 2 - np.abs(state[1]) ** 2
+
+        dev = qml.device("default.qubit", wires=1)
+        extra_param = np.array(0.31, requires_grad=False)
+
+        def cost(p1, p2):
+            with AutogradInterface.apply(ReversibleTape()) as tape:
+                qml.RX(extra_param, wires=[0])
+                qml.RY(p1, wires=[0])
+                qml.RZ(p2, wires=[0])
+                qml.RX(p1, wires=[0])
+                expval(qml.PauliZ(0))
+
+            assert tape.trainable_params == {1, 2, 3}
+            return tape.execute(dev)
+
+        zero_state = np.array([1.0, 0.0])
+
+        # analytic gradient
+        grad_fn = qml.grad(cost)
+        grad_A = grad_fn(reused_p, other_p)
+
+        # manual gradient
+        grad_true0 = (
+            expZ(
+                Rx(reused_p) @ Rz(other_p) @ Ry(reused_p + np.pi / 2) @ Rx(extra_param) @ zero_state
+            )
+            - expZ(
+                Rx(reused_p) @ Rz(other_p) @ Ry(reused_p - np.pi / 2) @ Rx(extra_param) @ zero_state
+            )
+        ) / 2
+        grad_true1 = (
+            expZ(
+                Rx(reused_p + np.pi / 2) @ Rz(other_p) @ Ry(reused_p) @ Rx(extra_param) @ zero_state
+            )
+            - expZ(
+                Rx(reused_p - np.pi / 2) @ Rz(other_p) @ Ry(reused_p) @ Rx(extra_param) @ zero_state
+            )
+        ) / 2
+        expected = grad_true0 + grad_true1  # product rule
+
+        assert np.allclose(grad_A[0], expected, atol=tol, rtol=0)
+
+    def test_gradient_gate_with_multiple_parameters(self, tol):
+        """Tests that gates with multiple free parameters yield correct gradients."""
+        x, y, z = [0.5, 0.3, -0.7]
+
+        with ReversibleTape() as tape:
+            qml.RX(0.4, wires=[0])
+            qml.Rot(x, y, z, wires=[0])
+            qml.RY(-0.2, wires=[0])
+            expval(qml.PauliZ(0))
+
+        tape.trainable_params = {1, 2, 3}
+
+        dev = qml.device("default.qubit", wires=1)
+        grad_A = tape.jacobian(dev, method="analytic")
+        grad_F = tape.jacobian(dev, method="numeric")
+
+        # gradient has the correct shape and every element is nonzero
+        assert grad_A.shape == (1, 3)
+        assert np.count_nonzero(grad_A) == 3
+        # the different methods agree
+        assert np.allclose(grad_A, grad_F, atol=tol, rtol=0)
+
+    def test_gradient_repeated_gate_parameters(self, mocker, tol):
+        """Tests that repeated use of a free parameter in a
+        multi-parameter gate yield correct gradients."""
+        dev = qml.device("default.qubit", wires=1)
+        params = np.array([0.8, 1.3], requires_grad=True)
+
+        def cost(params, method):
+            with AutogradInterface.apply(ReversibleTape()) as tape:

```

Comment:
Is it a matter of style?

E.g., would

```python
tape = ReversibleTape()
tape = AutogradInterface.apply(tape)

with tape:
```

look better?

Reply:
I tried to resist adding a `tape.to_autograd()` method (like we have with QNodes) as it always leads to issues with imports

Reply:
thinking something like `with ReversibleTape(interface="autograd")` like how we use interfaces everywhere else

Reply:
🤔 Could be doable. I think the reason I didn't do this is because I deliberately left the tape low-level (so it is a bit OOPy compared to the QNode). Thought it would be better to force you to explicitly import and apply the interface class, rather than have internal logic for it (like in the QNodes).

But if we think the tape might become more user-facing, should be easy enough to do